vispy/visuals/filters/clipping_planes.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+
+from functools import lru_cache
+
+import numpy as np
+
+from ..shaders import Function, Varying
+from .base_filter import Filter
+
+
+class PlanesClipper(Filter):
+    """Clips visual output based on arbitrary clipping planes."""
+
+    VERT_CODE = """
+    void clip() {
+        // Transform back to visual coordinates and clip based on that
```

Comment:
Shouldn't the clipping apply in world space? That is what ThreeJS does. It also makes is possible to use the same clipping planes for multiple visuals, regardless of their transform.

Reply:
We had some discussion about this in napari meetings (not sure if we have it written anywhere). What we realized there is that when interacting programmatically (via code, or for example with a slider), we intuitively thought in data coordinates (e.g: clip half of this volume, no matter where it is). However, live interactivity (as in the video example at the top of this conversation) feels intuitive in "world coordinates".

Ultimately, I think this depends a lot on the high level interface.

Reply:
I feel quite strongly that it should be world-coordinates, to be honest. If you want to apply the same clipping to multiple visuals in a scene, it feels to me that you should be able to apply the same planes. The data-space feels arbitrary. Also when think of the idea of moving data sources outside of the visuals.

Reply:
I'm ok with this change if you feels trongly about it. Anyways, I think that we should always want to interact directly with the visual (and therefore in world coords), because that feels better than with sliders and knobs.

I in case one wanted to, it should still be possible to recover the data-to-world transform and use that before giving the coordinates to vispy, right?

Reply:
I'm not sure I follow what you mean with interacting directly with the visual vs using sliders and knobs :)

>  it should still be possible to recover the data-to-world transform 

Yes, it should.

Reply:
I was trying to switch the clipping planes to world (scene?) coordinates, and I'm not sure how to proceed. When should the transform be applied, in your opinion? I guess the best computationally would be to do it in the setter, but I'm not sure if that's possible (I don't think I get the view's transforms there). Should this then happen in `_prepare_transforms`? Or should I just do it in the clipping plane function in the frag shader?

Reply:
Good question. I think it should happen in the fragment shader. You just upload the plane-definitions as they are and compare with the world coordinates in the shader. That's what ThreeJS does, and what we do in pygfx.

Reply:
@djhoese do you have any thoughts about using world coordinates (or do we call them scene-coordinates?) vs data coordinates for the planes?

Reply:
> do you have any thoughts 

No, not really :wink:. I think your point about wanted to use a single filter on multiple Visuals or potentially using this logic to apply to the entire Canvas at once makes sense for why it shouldn't be Visual-specific. I think it is unfortunate though that a user can't do the more logical data coordinates in the case where the filter is used on one object though. Would it make sense and be possible to make it configurable? :thinking: Or would it make sense to allow a transform to be added to the filter itself to make this logic easier? So in the default case it uses world coordinates, in the simple non-default case you could copy the Visual's `.transform` to this filter. :man_shrugging: Not sure it is worth the work but would be pretty cool.

Reply:
Being able to toggle makes sense to me. I just updated with a working version of this: any coordinate system can now be chosen for clipping for both the filter and volumes. Due to how transforms depend on `views` rather than `visuals`, I could not find a way to update the coordinate system on the fly; am I missing something?

Reply:
Being able to chose the space in to which the planes apply is awesome, nice work! So is this working now? Your last comment suggests you were still figuring part of it out...

Reply:
Sorry, wasn't clear. Yeah, it works! I was just pointing out that I had to implement this in a way that cannot be changed during visualization, due to the fact that setting transforms has to happen (afaik) before that (i.e: during `_attach` or `_prepare_transforms`)